File: python-chatdownloader/src/leaderboards/nonvips.py
"""
The non-VIPs leaderboard
"""

from typing import Optional

from _types import UserChatPerformance

from .abstractleaderboard import AbstractLeaderboard


class NonVIPS(AbstractLeaderboard):
    """
    Leaderboard for non-VIP users
    """
    @classmethod
    def get_name(cls):
        return 'nonvips'

    def calculate_score(self,
                        performance: UserChatPerformance) -> Optional[float]:
        if 'special_role' in performance.metadata \
           and performance.metadata['special_role']:
            return None
        return sum(performance.metrics.values())

File: rust/elo/src/leaderboards/nonvips.rs
/*
The non-VIPs leaderboard
*/

use crate::_types::clptypes::{MessageTag, UserChatPerformance};
use crate::_types::leaderboardtypes::LeaderboardInnerState;
use crate::is_message_origin;
use crate::leaderboards::leaderboardtrait::AbstractLeaderboard;
use std::collections::{HashMap, HashSet};

#[derive(Default, Debug)]
pub struct NonVIPS {
    state: HashMap<String, LeaderboardInnerState>,
}

impl AbstractLeaderboard for NonVIPS {
    fn new(optout_list: &HashSet<String>) -> Self {
        let mut out = Self {
            state: HashMap::new(),
        };
        out.read_initial_state();
        out.cull_optout(optout_list);
        out
    }

    fn get_name(&self) -> String {
        "nonvips".to_string()
    }

    fn __get_state(&mut self) -> &mut HashMap<String, LeaderboardInnerState> {
        &mut self.state
    }

    fn calculate_score(&self, performance: &UserChatPerformance) -> Option<f32> {
        if is_message_origin!(performance, MessageTag::Discord) {
            return None;
        }

        if let Some(special_role) = performance.metadata.get("special_role") {
            if *special_role.get_bool().unwrap_or(&false) {
                return None;
            }
        }
        Some(performance.metrics.values().sum())
    }
}
